### PR TITLE
refactor(connector): [KLARNA] Add dynamic fields for klarna payment method

### PR DESCRIPTION
**Pull request by swangi-kumari**
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [x] Enhancement
- [x] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description
Add dynamic fields for klarna payment method

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?
1. Create a Klarna Payment Request via Klarna with with `confirm=fasle`
```
{
    "amount": 1800,
    "currency": "USD",
    "confirm": false,
    "capture_method": "automatic",
    "customer_id": "cus_NwicjmCJbgafjhjfhHr6",
    "capture_on": "2022-09-10T10:11:12Z",
    "authentication_type": "three_ds",
    "email": "guest@example.com",
    "name": "John Doe",
    "phone": "9123456789",
    "phone_country_code": "+65",
    "description": "Its my first payment request",
    "return_url": "https://duck.com",
    "statement_descriptor_name": "Juspay",
    "statement_descriptor_suffix": "Router",
    "payment_method": "pay_later",
    "payment_method_type": "klarna",
    "payment_experience": "invoke_sdk_client",
    "payment_method_data": {
        "pay_later": {
            "klarna_sdk": {
                "token": "65bfb654-ef2c-6b39-942e-ffdc9a105d3cf"
            }
        }
    },
    "billing": {
        "address": {
            "line1": "444, Stelar Enclave",
            "line2": "California",
            "line3": "Groningen",
            "city": "Groningen",
            "state": "Groningen",
            "zip": "94105",
            "country": "US",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "+4901761428434",
            "country_code": "+1"
        }
    },
    "shipping": {
        "address": {
            "line1": "444, Stelar Enclave",
            "line2": "California",
            "line3": "Groningen",
            "city": "Groningen",
            "state": "Groningen",
            "zip": "94105",
            "country": "US",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "446681800",
            "country_code": "+41"
        },
        "email": "guest@example.com"
    },
    "metadata": {},
    "order_details": [
        {
            "product_name": "Apple iphone 15",
            "quantity": 2,
            "amount": 1800,
            "account_name": "transaction_processing"
        }
    ],
    "browser_info": {
        "user_agent": "Mozilla\/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit\/537.36 (KHTML, like Gecko) Chrome\/70.0.3538.110 Safari\/537.36",
        "accept_header": "text\/html,application\/xhtml+xml,application\/xml;q=0.9,image\/webp,image\/apng,*\/*;q=0.8",
        "language": "nl-NL",
        "color_depth": 24,
        "screen_height": 723,
        "screen_width": 1536,
        "time_zone": 0,
        "java_enabled": true,
        "java_script_enabled": true,
        "ip_address": "128.0.0.1"
    }
}
```
Response:
```
{
    "payment_id": "pay_BtRF93AhE0erz2yHuryN",
    "merchant_id": "merchant_1717586050",
    "status": "requires_confirmation",
    "amount": 1800,
    "net_amount": 1800,
    "amount_capturable": 0,
    "amount_received": null,
    "connector": null,
    "client_secret": "pay_BtRF93AhE0erz2yHuryN_secret_fzSL7QM3OB7TwlfKH5jC",
    "created": "2024-06-05T13:23:40.457Z",
    "currency": "USD",
    "customer_id": "cus_NwicjmCJbgafjhjfhHr6",
    "customer": {
        "id": "cus_NwicjmCJbgafjhjfhHr6",
        "name": "John Doe",
        "email": "guest@example.com",
        "phone": "9123456789",
        "phone_country_code": "+65"
    },
    "description": "Its my first payment request",
    "refunds": null,
    "disputes": null,
    "mandate_id": null,
    "mandate_data": null,
    "setup_future_usage": null,
    "off_session": null,
    "capture_on": null,
    "capture_method": "automatic",
    "payment_method": "pay_later",
    "payment_method_data": {
        "pay_later": {},
        "billing": null
    },
    "payment_token": null,
    "shipping": {
        "address": {
            "city": "Groningen",
            "country": "US",
            "line1": "444, Stelar Enclave",
            "line2": "California",
            "line3": "Groningen",
            "zip": "94105",
            "state": "Groningen",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "446681800",
            "country_code": "+41"
        },
        "email": "guest@example.com"
    },
    "billing": {
        "address": {
            "city": "Groningen",
            "country": "US",
            "line1": "444, Stelar Enclave",
            "line2": "California",
            "line3": "Groningen",
            "zip": "94105",
            "state": "Groningen",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "+4901761428434",
            "country_code": "+1"
        },
        "email": null
    },
    "order_details": [
        {
            "brand": null,
            "amount": 1800,
            "category": null,
            "quantity": 2,
            "product_id": null,
            "product_name": "Apple iphone 15",
            "product_type": null,
            "sub_category": null,
            "product_img_link": null,
            "requires_shipping": null
        }
    ],
    "email": "guest@example.com",
    "name": "John Doe",
    "phone": "9123456789",
    "return_url": "https://duck.com/",
    "authentication_type": "three_ds",
    "statement_descriptor_name": "Juspay",
    "statement_descriptor_suffix": "Router",
    "next_action": null,
    "cancellation_reason": null,
    "error_code": null,
    "error_message": null,
    "unified_code": null,
    "unified_message": null,
    "payment_experience": "invoke_sdk_client",
    "payment_method_type": "klarna",
    "connector_label": null,
    "business_country": null,
    "business_label": "default",
    "business_sub_label": null,
    "allowed_payment_method_types": null,
    "ephemeral_key": {
        "customer_id": "cus_NwicjmCJbgafjhjfhHr6",
        "created_at": 1717593820,
        "expires": 1717597420,
        "secret": "epk_83b98d86e60a439e9cdf5e3ee3007e84"
    },
    "manual_retry_allowed": null,
    "connector_transaction_id": null,
    "frm_message": null,
    "metadata": {},
    "connector_metadata": null,
    "feature_metadata": null,
    "reference_id": null,
    "payment_link": null,
    "profile_id": "pro_eVJTR1hn3FYTrnYncTdW",
    "surcharge_details": null,
    "attempt_count": 1,
    "merchant_decision": null,
    "merchant_connector_id": null,
    "incremental_authorization_allowed": null,
    "authorization_count": null,
    "incremental_authorizations": null,
    "external_authentication_details": null,
    "external_3ds_authentication_attempted": false,
    "expires_on": "2024-06-05T13:38:40.457Z",
    "fingerprint": null,
    "browser_info": {
        "language": "nl-NL",
        "time_zone": 0,
        "ip_address": "128.0.0.1",
        "user_agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/70.0.3538.110 Safari/537.36",
        "color_depth": 24,
        "java_enabled": true,
        "screen_width": 1536,
        "accept_header": "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8",
        "screen_height": 723,
        "java_script_enabled": true
    },
    "payment_method_id": null,
    "payment_method_status": null,
    "updated": "2024-06-05T13:23:40.489Z",
    "charges": null,
    "frm_metadata": null
}
```
____________________________________________________

Run this DB query to  collect shipping_address
```
update business_profile set collect_shipping_details_from_wallet_connector = true where profile_id ='{profile_id}';
```

2. List Payment Methods for a Merchant
```
curl --location 'http://localhost:8080/account/payment_methods?client_secret=pay_BtRF93AhE0erz2yHuryN_secret_fzSL7QM3OB7TwlfKH5jC' \
--header 'Accept: application/json' \
--header 'api-key: {api_key}'
```
Response:
```
{
    "redirect_url": "https://google.com/success",
    "currency": "USD",
    "payment_methods": [
        {
            "payment_method": "pay_later",
            "payment_method_types": [
                {
                    "payment_method_type": "klarna",
                    "payment_experience": [
                        {
                            "payment_experience_type": "invoke_sdk_client",
                            "eligible_connectors": [
                                "klarna"
                            ]
                        },
                        {
                            "payment_experience_type": "redirect_to_url",
                            "eligible_connectors": [
                                "klarna"
                            ]
                        }
                    ],
                    "card_networks": null,
                    "bank_names": null,
                    "bank_debits": null,
                    "bank_transfers": null,
                    "required_fields": {
                        "shipping.address.line2": {
                            "required_field": "shipping.address.line2",
                            "display_name": "line2",
                            "field_type": "user_shipping_address_line2",
                            "value": "California"
                        },
                        "shipping.phone.country_code": {
                            "required_field": "shipping.phone.country_code",
                            "display_name": "phone_country_code",
                            "field_type": "user_phone_number_country_code",
                            "value": "+41"
                        },
                        "shipping.address.line1": {
                            "required_field": "shipping.address.line1",
                            "display_name": "line1",
                            "field_type": "user_shipping_address_line1",
                            "value": "444, Stelar Enclave"
                        },
                        "shipping.email": {
                            "required_field": "shipping.email",
                            "display_name": "email",
                            "field_type": "user_email_address",
                            "value": "guest@example.com"
                        },
                        "shipping.address.first_name": {
                            "required_field": "shipping.address.first_name",
                            "display_name": "shipping_first_name",
                            "field_type": "user_shipping_name",
                            "value": "joseph"
                        },
                        "shipping.address.country": {
                            "required_field": "shipping.address.country",
                            "display_name": "country",
                            "field_type": {
                                "user_shipping_address_country": {
                                    "options": [
                                        "AU",
                                        "AT",
                                        "BE",
                                        "CA",
                                        "CZ",
                                        "DK",
                                        "FI",
                                        "FR",
                                        "DE",
                                        "GR",
                                        "IE",
                                        "IT",
                                        "NL",
                                        "NZ",
                                        "NO",
                                        "PL",
                                        "PT",
                                        "ES",
                                        "SE",
                                        "CH",
                                        "GB",
                                        "US"
                                    ]
                                }
                            },
                            "value": "US"
                        },
                        "shipping.address.state": {
                            "required_field": "shipping.address.state",
                            "display_name": "state",
                            "field_type": "user_shipping_address_state",
                            "value": "Groningen"
                        },
                        "shipping.address.zip": {
                            "required_field": "shipping.address.zip",
                            "display_name": "zip",
                            "field_type": "user_shipping_address_pincode",
                            "value": "94105"
                        },
                        "shipping.phone.number": {
                            "required_field": "shipping.phone.number",
                            "display_name": "phone_number",
                            "field_type": "user_phone_number",
                            "value": "446681800"
                        },
                        "shipping.address.last_name": {
                            "required_field": "shipping.address.last_name",
                            "display_name": "shipping_last_name",
                            "field_type": "user_shipping_name",
                            "value": "Doe"
                        },
                        "shipping.address.city": {
                            "required_field": "shipping.address.city",
                            "display_name": "city",
                            "field_type": "user_shipping_address_city",
                            "value": "Groningen"
                        },
                    },
                    "surcharge_details": null,
                    "pm_auth_connector": null
                }
            ]
        }
    ],
    "mandate_payment": null,
    "merchant_name": "NewAge Retailer",
    "show_surcharge_breakup_screen": false,
    "payment_type": "normal",
    "request_external_three_ds_authentication": false
}
```

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

File: crates/router/src/configs/defaults.rs [hunk 1]
```diff
@@ -8239,6 +8239,139 @@ impl Default for super::settings::RequiredFields {
                                         common : HashMap::new(),
                                     }
                                 ),
+                                (
+                                    enums::Connector::Klarna,
+                                    RequiredFieldFinal {
+                                        mandate: HashMap::new(),
+                                        non_mandate: HashMap::from([
```
Comment (SamraatBansal) [inline crates/router/src/configs/defaults.rs:8246]: Should be in `common` field

File: crates/router/src/configs/defaults.rs [hunk 2]
```diff
@@ -8239,6 +8239,139 @@ impl Default for super::settings::RequiredFields {
...
+                                            (
+                                                "shipping.address.first_name".to_string(),
+                                                RequiredFieldInfo {
+                                                    required_field: "shipping.address.first_name".to_string(),
+                                                    display_name: "given_name".to_string(),
```
Comment (SamraatBansal) [inline crates/router/src/configs/defaults.rs:8251]: We need to use our terminology here, so whatever is used as display_name for this field across SDK Should be consistent

Comment (SamraatBansal) [reply crates/router/src/configs/defaults.rs:8251]: Please do this at all the required places

Comment (SamraatBansal) [reply crates/router/src/configs/defaults.rs:8251]: Also confirm from SDK do they need this field in Express checkout flow at someplace.

**Review by SamraatBansal — CHANGES_REQUESTED** [review]

File: crates/router/src/configs/defaults.rs [hunk 3]
```diff
@@ -8239,6 +8239,139 @@ impl Default for super::settings::RequiredFields {
...
+                                                    required_field: "shipping.email".to_string(),
+                                                    display_name: "email".to_string(),
+                                                    field_type: enums::FieldType::UserEmailAddress,
+                                                    value: None,
+                                                }
+                                            ),
+                                            (
+                                                "shipping.phone.number".to_string(),
+                                                RequiredFieldInfo {
+                                                    required_field: "payment_method_data.shipping.phone.number".to_string(),
```
Comment (SamraatBansal) [inline crates/router/src/configs/defaults.rs:8323]:
```suggestion
                                                    required_field: "shipping.phone.number".to_string(),
```

File: crates/router/src/configs/defaults.rs [hunk 4]
```diff
@@ -8239,6 +8239,139 @@ impl Default for super::settings::RequiredFields {
...
+                                                    display_name: "phone".to_string(),
```
Comment (SamraatBansal) [inline crates/router/src/configs/defaults.rs:8324]:
```suggestion
                                                    display_name: "phone_number".to_string(),
```

File: crates/router/src/configs/defaults.rs [hunk 5]
```diff
@@ -8239,6 +8239,139 @@ impl Default for super::settings::RequiredFields {
...
+                                                    field_type: enums::FieldType::UserPhoneNumber,
+                                                    value: None,
+                                                }
+                                            ),
+                                            (
+                                                "shipping.phone.country_code".to_string(),
+                                                RequiredFieldInfo {
+                                                    required_field: "payment_method_data.shipping.phone.country_code".to_string(),
```
Comment (SamraatBansal) [inline crates/router/src/configs/defaults.rs:8332]:
```suggestion
                                                    required_field: "shipping.phone.country_code".to_string(),
```

File: crates/router/src/configs/defaults.rs [hunk 6]
```diff
@@ -8239,6 +8239,139 @@ impl Default for super::settings::RequiredFields {
...
+                                                    display_name: "phone".to_string(),
```
Comment (SamraatBansal) [inline crates/router/src/configs/defaults.rs:8333]:
```suggestion
                                                    display_name: "phone_country_code".to_string(),
```

**Review by SamraatBansal — CHANGES_REQUESTED** [review]

**Review by SamraatBansal — DISMISSED** [review]

**Review by sai-harsha-vardhan — APPROVED** [review]

**Review by srujanchikke — DISMISSED** [review]

**Review by SamraatBansal — APPROVED** [review]

**Review by sai-harsha-vardhan — APPROVED** [review]

**Pull request merged**